{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "yjdblManpnnM"
   },
   "source": [
    "# PDF Acquisition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "bYq4tu_roBQg"
   },
   "outputs": [],
   "source": [
    "#@title Depandances\n",
    "%%capture\n",
    "!pip3 install -U scidownl\n",
    "# Langchain and groq\n",
    "!pip install langchain\n",
    "!pip install langchain_groq\n",
    "\n",
    "#installing marker\n",
    "!pip install marker-pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 342
    },
    "id": "BdNn168OoOwq",
    "outputId": "db1726c3-2b6e-4636-8101-c7550876da47"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "     <input type=\"file\" id=\"files-1cf9622b-a058-4ff4-9ade-346c2e4185f9\" name=\"files[]\" multiple disabled\n",
       "        style=\"border:none\" />\n",
       "     <output id=\"result-1cf9622b-a058-4ff4-9ade-346c2e4185f9\">\n",
       "      Upload widget is only available when the cell has been executed in the\n",
       "      current browser session. Please rerun this cell to enable.\n",
       "      </output>\n",
       "      <script>// Copyright 2017 Google LLC\n",
       "//\n",
       "// Licensed under the Apache License, Version 2.0 (the \"License\");\n",
       "// you may not use this file except in compliance with the License.\n",
       "// You may obtain a copy of the License at\n",
       "//\n",
       "//      http://www.apache.org/licenses/LICENSE-2.0\n",
       "//\n",
       "// Unless required by applicable law or agreed to in writing, software\n",
       "// distributed under the License is distributed on an \"AS IS\" BASIS,\n",
       "// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n",
       "// See the License for the specific language governing permissions and\n",
       "// limitations under the License.\n",
       "\n",
       "/**\n",
       " * @fileoverview Helpers for google.colab Python module.\n",
       " */\n",
       "(function(scope) {\n",
       "function span(text, styleAttributes = {}) {\n",
       "  const element = document.createElement('span');\n",
       "  element.textContent = text;\n",
       "  for (const key of Object.keys(styleAttributes)) {\n",
       "    element.style[key] = styleAttributes[key];\n",
       "  }\n",
       "  return element;\n",
       "}\n",
       "\n",
       "// Max number of bytes which will be uploaded at a time.\n",
       "const MAX_PAYLOAD_SIZE = 100 * 1024;\n",
       "\n",
       "function _uploadFiles(inputId, outputId) {\n",
       "  const steps = uploadFilesStep(inputId, outputId);\n",
       "  const outputElement = document.getElementById(outputId);\n",
       "  // Cache steps on the outputElement to make it available for the next call\n",
       "  // to uploadFilesContinue from Python.\n",
       "  outputElement.steps = steps;\n",
       "\n",
       "  return _uploadFilesContinue(outputId);\n",
       "}\n",
       "\n",
       "// This is roughly an async generator (not supported in the browser yet),\n",
       "// where there are multiple asynchronous steps and the Python side is going\n",
       "// to poll for completion of each step.\n",
       "// This uses a Promise to block the python side on completion of each step,\n",
       "// then passes the result of the previous step as the input to the next step.\n",
       "function _uploadFilesContinue(outputId) {\n",
       "  const outputElement = document.getElementById(outputId);\n",
       "  const steps = outputElement.steps;\n",
       "\n",
       "  const next = steps.next(outputElement.lastPromiseValue);\n",
       "  return Promise.resolve(next.value.promise).then((value) => {\n",
       "    // Cache the last promise value to make it available to the next\n",
       "    // step of the generator.\n",
       "    outputElement.lastPromiseValue = value;\n",
       "    return next.value.response;\n",
       "  });\n",
       "}\n",
       "\n",
       "/**\n",
       " * Generator function which is called between each async step of the upload\n",
       " * process.\n",
       " * @param {string} inputId Element ID of the input file picker element.\n",
       " * @param {string} outputId Element ID of the output display.\n",
       " * @return {!Iterable<!Object>} Iterable of next steps.\n",
       " */\n",
       "function* uploadFilesStep(inputId, outputId) {\n",
       "  const inputElement = document.getElementById(inputId);\n",
       "  inputElement.disabled = false;\n",
       "\n",
       "  const outputElement = document.getElementById(outputId);\n",
       "  outputElement.innerHTML = '';\n",
       "\n",
       "  const pickedPromise = new Promise((resolve) => {\n",
       "    inputElement.addEventListener('change', (e) => {\n",
       "      resolve(e.target.files);\n",
       "    });\n",
       "  });\n",
       "\n",
       "  const cancel = document.createElement('button');\n",
       "  inputElement.parentElement.appendChild(cancel);\n",
       "  cancel.textContent = 'Cancel upload';\n",
       "  const cancelPromise = new Promise((resolve) => {\n",
       "    cancel.onclick = () => {\n",
       "      resolve(null);\n",
       "    };\n",
       "  });\n",
       "\n",
       "  // Wait for the user to pick the files.\n",
       "  const files = yield {\n",
       "    promise: Promise.race([pickedPromise, cancelPromise]),\n",
       "    response: {\n",
       "      action: 'starting',\n",
       "    }\n",
       "  };\n",
       "\n",
       "  cancel.remove();\n",
       "\n",
       "  // Disable the input element since further picks are not allowed.\n",
       "  inputElement.disabled = true;\n",
       "\n",
       "  if (!files) {\n",
       "    return {\n",
       "      response: {\n",
       "        action: 'complete',\n",
       "      }\n",
       "    };\n",
       "  }\n",
       "\n",
       "  for (const file of files) {\n",
       "    const li = document.createElement('li');\n",
       "    li.append(span(file.name, {fontWeight: 'bold'}));\n",
       "    li.append(span(\n",
       "        `(${file.type || 'n/a'}) - ${file.size} bytes, ` +\n",
       "        `last modified: ${\n",
       "            file.lastModifiedDate ? file.lastModifiedDate.toLocaleDateString() :\n",
       "                                    'n/a'} - `));\n",
       "    const percent = span('0% done');\n",
       "    li.appendChild(percent);\n",
       "\n",
       "    outputElement.appendChild(li);\n",
       "\n",
       "    const fileDataPromise = new Promise((resolve) => {\n",
       "      const reader = new FileReader();\n",
       "      reader.onload = (e) => {\n",
       "        resolve(e.target.result);\n",
       "      };\n",
       "      reader.readAsArrayBuffer(file);\n",
       "    });\n",
       "    // Wait for the data to be ready.\n",
       "    let fileData = yield {\n",
       "      promise: fileDataPromise,\n",
       "      response: {\n",
       "        action: 'continue',\n",
       "      }\n",
       "    };\n",
       "\n",
       "    // Use a chunked sending to avoid message size limits. See b/62115660.\n",
       "    let position = 0;\n",
       "    do {\n",
       "      const length = Math.min(fileData.byteLength - position, MAX_PAYLOAD_SIZE);\n",
       "      const chunk = new Uint8Array(fileData, position, length);\n",
       "      position += length;\n",
       "\n",
       "      const base64 = btoa(String.fromCharCode.apply(null, chunk));\n",
       "      yield {\n",
       "        response: {\n",
       "          action: 'append',\n",
       "          file: file.name,\n",
       "          data: base64,\n",
       "        },\n",
       "      };\n",
       "\n",
       "      let percentDone = fileData.byteLength === 0 ?\n",
       "          100 :\n",
       "          Math.round((position / fileData.byteLength) * 100);\n",
       "      percent.textContent = `${percentDone}% done`;\n",
       "\n",
       "    } while (position < fileData.byteLength);\n",
       "  }\n",
       "\n",
       "  // All done.\n",
       "  yield {\n",
       "    response: {\n",
       "      action: 'complete',\n",
       "    }\n",
       "  };\n",
       "}\n",
       "\n",
       "scope.google = scope.google || {};\n",
       "scope.google.colab = scope.google.colab || {};\n",
       "scope.google.colab._files = {\n",
       "  _uploadFiles,\n",
       "  _uploadFilesContinue,\n",
       "};\n",
       "})(self);\n",
       "</script> "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-fe3794aac9db>\u001b[0m in \u001b[0;36m<cell line: 32>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     31\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m__name__\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"__main__\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 33\u001b[0;31m     \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-4-fe3794aac9db>\u001b[0m in \u001b[0;36mmain\u001b[0;34m()\u001b[0m\n\u001b[1;32m     26\u001b[0m         \u001b[0mdownload_paper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mchoice\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'2'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 28\u001b[0;31m         \u001b[0mupload_paper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     29\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     30\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Invalid choice. Please enter 1 or 2.\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-4-fe3794aac9db>\u001b[0m in \u001b[0;36mupload_paper\u001b[0;34m()\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mupload_paper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 17\u001b[0;31m     \u001b[0muploaded\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfiles\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     18\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muploaded\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mendswith\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'.pdf'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Uploaded file is not a PDF file.'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/files.py\u001b[0m in \u001b[0;36mupload\u001b[0;34m()\u001b[0m\n\u001b[1;32m     67\u001b[0m   \"\"\"\n\u001b[1;32m     68\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 69\u001b[0;31m   \u001b[0muploaded_files\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_upload_files\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmultiple\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     70\u001b[0m   \u001b[0;31m# Mapping from original filename to filename as saved locally.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     71\u001b[0m   \u001b[0mlocal_filenames\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/files.py\u001b[0m in \u001b[0;36m_upload_files\u001b[0;34m(multiple)\u001b[0m\n\u001b[1;32m    162\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    163\u001b[0m   \u001b[0;32mwhile\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'action'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'complete'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 164\u001b[0;31m     result = _output.eval_js(\n\u001b[0m\u001b[1;32m    165\u001b[0m         'google.colab._files._uploadFilesContinue(\"{output_id}\")'.format(\n\u001b[1;32m    166\u001b[0m             \u001b[0moutput_id\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0moutput_id\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/output/_js.py\u001b[0m in \u001b[0;36meval_js\u001b[0;34m(script, ignore_result, timeout_sec)\u001b[0m\n\u001b[1;32m     38\u001b[0m   \u001b[0;32mif\u001b[0m \u001b[0mignore_result\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0;32mreturn\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m   \u001b[0;32mreturn\u001b[0m \u001b[0m_message\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_reply_from_input\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout_sec\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/google/colab/_message.py\u001b[0m in \u001b[0;36mread_reply_from_input\u001b[0;34m(message_id, timeout_sec)\u001b[0m\n\u001b[1;32m     94\u001b[0m     \u001b[0mreply\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_read_next_input_message\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     95\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mreply\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0m_NOT_READY\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreply\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 96\u001b[0;31m       \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msleep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0.025\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     97\u001b[0m       \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     98\u001b[0m     if (\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from scidownl import scihub_download\n",
    "import pandas as pd\n",
    "import os\n",
    "from google.colab import files\n",
    "\n",
    "def download_paper():\n",
    "    paper = input(\"Please enter the DOI of the paper here: \")\n",
    "    paper_type = \"doi\"\n",
    "    out = \"./paper.pdf\"\n",
    "    # Uncomment and configure the proxies if needed\n",
    "    # proxies = {\n",
    "    #     'http': 'socks5://127.0.0.1:7890'\n",
    "    # }\n",
    "    scihub_download(paper, paper_type=paper_type, out=out)#, proxies=proxies)\n",
    "\n",
    "def upload_paper():\n",
    "    uploaded = files.upload()\n",
    "    if not list(uploaded.keys())[0].endswith('.pdf'):\n",
    "        raise ValueError('Uploaded file is not a PDF file.')\n",
    "    os.rename(list(uploaded.keys())[0], 'paper.pdf')\n",
    "    print(\"Paper uploaded successfully as 'paper.pdf'\")\n",
    "\n",
    "def main():\n",
    "    choice = input(\"Choose an option (1 for DOI download, 2 for PDF upload): \")\n",
    "    if choice == '1':\n",
    "        download_paper()\n",
    "    elif choice == '2':\n",
    "        upload_paper()\n",
    "    else:\n",
    "        print(\"Invalid choice. Please enter 1 or 2.\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "u871l3MWPQ7c"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "I7w_hmCOpvnd"
   },
   "source": [
    "# Markdown Extraction with Marker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "TlzWKPpHUHNL"
   },
   "outputs": [],
   "source": [
    "#@title extracting markdown\n",
    "!marker_single /content/paper.pdf /content/output/ --batch_multiplier 2 --max_pages 20 --langs English"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "RogGmG7cp8-e"
   },
   "source": [
    "# Extracting the Methodology Section\n",
    "**Objective:** Isolate the methodology section from the segmented markdown for further analysis.\n",
    "- **Action Items:**\n",
    "  - From the parsed sections, specifically extract the text under the \"Methodology\" heading.\n",
    "  - Handle variations in heading titles like \"Methods\" or \"Materials and Methods.\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "QGyWoVDrp-A0"
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def extract_outline(file_path):\n",
    "    outline = []\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            header_match = re.match(r'^(#+)\\s+(.*)', line)\n",
    "            if header_match:\n",
    "                level = len(header_match.group(1))\n",
    "                title = header_match.group(2)\n",
    "                outline.append((header_match.group(1), title))\n",
    "    return outline\n",
    "\n",
    "def print_outline(outline):\n",
    "    for hashes, title in outline:\n",
    "        print(hashes + ' ' + title)\n",
    "\n",
    "file_path = 'manuscript.md'\n",
    "outline = extract_outline(file_path)\n",
    "print_outline(outline)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "id": "NN93TF30WvUV"
   },
   "outputs": [],
   "source": [
    "from langchain_google_genai import ChatGoogleGenerativeAI\n",
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "\n",
    "from prompts import system_prompt_1, system_prompt_2, get_markdown\n",
    "\n",
    "import re\n",
    "import os \n",
    "from IPython.display import Markdown\n",
    "\n",
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "from langchain_groq import ChatGroq\n",
    "\n",
    "\n",
    "\n",
    "# setup google gemini api keys\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "GOOGLE_API_KEY = os.getenv(\"GOOGLE_API_KEY\")\n",
    "\n",
    "gem1 = ChatGoogleGenerativeAI(model=\"gemini-1.5-pro\", google_api_key=GOOGLE_API_KEY, temperature=0.4, convert_system_message_to_human=True)\n",
    "flash = ChatGoogleGenerativeAI(model=\"gemini-1.5-flash\", google_api_key=GOOGLE_API_KEY, temperature=0.3, convert_system_message_to_human=True)\n",
    "\n",
    "\n",
    "GROQ_API_KEY = os.getenv(\"GROQ_API_KEY\")\n",
    "\n",
    "groq_mx = ChatGroq(\n",
    "    temperature=0,\n",
    "    # model_name=\"mixtral-8x7b-32768\",\n",
    "    # model_name=\"llama3-70b-8192\",\n",
    "    # model_name=\"llama-3.1-70b-versatile\",\n",
    "    # model_name=\"mixtral-8x7b-32768\",\n",
    "    # model_name=\"llama3-8b-8192\",\n",
    "    model_name=\"llama-3.1-8b-instant\",\n",
    "    # model_name=\"gemma2-9b-it\",\n",
    "    \n",
    "    groq_api_key=GROQ_API_KEY \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "o0RlXtxQTiiM",
    "outputId": "7e5d35a0-a54f-4a3a-a06c-06d50b3c9e77"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Chapter Two: Materials and Methods\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "system = \"\"\"\n",
    "###instructions###\n",
    "You will receive an outline of a scientific paper in the format shown in the expected input. Extract and present only the highest level heading corresponding to the methods section, with no other output.\n",
    "\n",
    "**Expected Input:**\n",
    "#abstract\n",
    "#introduction\n",
    "#method\n",
    "##method 1\n",
    "#results\n",
    "#conclusion\n",
    "\n",
    "**Expected Output:**\n",
    "#method\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "text = f\"\"\"{outline}\"\"\"\n",
    "prompt = ChatPromptTemplate.from_messages([(\"system\", system), (\"human\", text)])\n",
    "\n",
    "chain = prompt | chat\n",
    "response = chain.invoke({})\n",
    "print(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "cellView": "form",
    "id": "kkisie-yTgs2"
   },
   "outputs": [],
   "source": [
    "#@title Extracting the method section from the Markdown file\n",
    "import re\n",
    "\n",
    "def extract_section(file_path, section_title):\n",
    "    section_content = []\n",
    "    inside_section = False\n",
    "    section_level = None\n",
    "\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            header_match = re.match(r'^(#+)\\s+(.*)', line)\n",
    "            if header_match:\n",
    "                level = len(header_match.group(1))\n",
    "                title = header_match.group(2)\n",
    "\n",
    "                if inside_section:\n",
    "                    if level <= section_level:\n",
    "                        break\n",
    "                if header_match.group(1) + ' ' + title == section_title:\n",
    "                    inside_section = True\n",
    "                    section_level = level\n",
    "\n",
    "            if inside_section:\n",
    "                section_content.append(line)\n",
    "\n",
    "    return ''.join(section_content)\n",
    "\n",
    "\n",
    "file_path = 'manuscript.md'\n",
    "section_title = response.content\n",
    "method = extract_section(file_path, section_title)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "7v_dig2-h2BD"
   },
   "source": [
    "## Comperhencive Sammary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "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\n"
     ]
    }
   ],
   "source": [
    "from dotenv import load_dotenv\n",
    "import os\n",
    "load_dotenv()\n",
    "GOOGLE_API_KEY = os.getenv(\"GOOGLE_API_KEY\")\n",
    "\n",
    "\n",
    "GROQ_API_KEY = os.getenv(\"GROQ_API_KEY\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_core.prompts import ChatPromptTemplate\n",
    "from langchain_groq import ChatGroq\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from prompts import system_prompt_1, system_prompt_2\n",
    "from input import method_1, method_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "cellView": "form",
    "id": "z0DJ2K0fqCUl"
   },
   "outputs": [],
   "source": [
    "#@title creation of nodes and edges\n",
    "\n",
    "chat = ChatGroq(\n",
    "    temperature=0,\n",
    "    model_name=\"mixtral-8x7b-32768\",\n",
    "    \n",
    "    max_tokens=1024,\n",
    "    groq_api_key=GROQ_API_KEY  # Optional if not set as an environment variable\n",
    ")\n",
    "\n",
    "system = system_prompt_1\n",
    "\n",
    "text = f\"\"\"{method_2}\"\"\"\n",
    "prompt = ChatPromptTemplate.from_messages([(\"system\", system), (\"human\", text)])\n",
    "\n",
    "chain = prompt | chat\n",
    "nande = chain.invoke({})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "cellView": "form",
    "id": "_6LZcAfcGz6S"
   },
   "outputs": [],
   "source": [
    "from prompts import system_prompt_1, system_prompt_2\n",
    "from input import method_1, method_2\n",
    "\n",
    "#@title writing the code for the graphviz\n",
    "\n",
    "chat = ChatGroq(\n",
    "    temperature=0,\n",
    "    model_name=\"gemma2-9b-it\",\n",
    "    max_tokens=1024,\n",
    "    groq_api_key=GROQ_API_KEY  # Optional if not set as an environment variable\n",
    ")\n",
    "\n",
    "system = system_prompt_2\n",
    "\n",
    "text = f\"\"\"{nande.content}\"\"\"\n",
    "prompt = ChatPromptTemplate.from_messages([(\"system\", system), (\"human\", text)])\n",
    "\n",
    "chain = prompt | chat\n",
    "llm_response = chain.invoke({})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "cellView": "form",
    "id": "uOKa35CqIEJt"
   },
   "outputs": [],
   "source": [
    "#@title extract the code form the response and run it\n",
    "import re\n",
    "\n",
    "def extract_code(llm_response):\n",
    "    # Define the pattern to match the Python code block\n",
    "    pattern = re.compile(r'```python(.*?)```', re.DOTALL)\n",
    "\n",
    "    # Search for the code block\n",
    "    match = pattern.search(llm_response)\n",
    "\n",
    "    if match:\n",
    "        # Extract the code\n",
    "        code = match.group(1).strip()\n",
    "        return code\n",
    "    else:\n",
    "        return \"No Python code found in the response.\"\n",
    "\n",
    "# Example LLM response\n",
    "code = extract_code(llm_response.content)\n",
    "exec(code)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "### instructions ###\n",
      "Your job is to write Python code that:\n",
      "1. Reads the provided list of nodes and edges.\n",
      "2. Utilizes Graphviz to create a directed graph.\n",
      "3. Adds nodes to the graph with appropriate labels.\n",
      "4. Adds edges between the nodes with descriptions of their relationships.\n",
      "5. Sets aesthetic properties such as layout direction, size, shape, or other visual attributes.\n",
      "7. Very important to output the results to high quality png image name \"graph.png\". \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(system_prompt_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comperhencive sammary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Manuscript to pandas\n",
    "import util\n",
    "import importlib\n",
    "importlib.reload(util)\n",
    "markdown_path = 'Examples/systematic_review/systematic_review.md'\n",
    "\n",
    "df = util.create_dataframe_from_sections(markdown_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "llm = ChatGroq(\n",
    "    temperature=0,\n",
    "    # model_name=\"mixtral-8x7b-32768\",\n",
    "    # model_name=\"llama3-70b-8192\",\n",
    "    # model_name=\"llama-3.1-70b-versatile\",\n",
    "    # model_name=\"mixtral-8x7b-32768\",\n",
    "    model_name=\"llama3-8b-8192\",\n",
    "    # model_name=\"llama-3.1-8b-instant\",\n",
    "    # model_name=\"gemma2-9b-it\",\n",
    "    \n",
    "    groq_api_key=GROQ_API_KEY \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_summary(content):\n",
    "    system = \"\"\"\n",
    "###instructions###\n",
    "You are a sammarizer you job is to sammarize any input into a structured markdown sammary that is comberhencive.\n",
    "\n",
    "## don't include a conclusion section\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "    prompt = ChatPromptTemplate.from_messages([(\"system\", system), (\"human\", content)])\n",
    "    \n",
    "    chain = prompt | llm\n",
    "    response = chain.invoke({})\n",
    "    return response.content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "BadRequestError",
     "evalue": "Error code: 400 - {'error': {'message': 'Please reduce the length of the messages or completion.', 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mBadRequestError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[13], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;66;03m# Add the summary column using a for loop\u001b[39;00m\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m index, row \u001b[38;5;129;01min\u001b[39;00m df\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[1;32m----> 3\u001b[0m     df\u001b[38;5;241m.\u001b[39mat[index, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124msummary\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m generate_summary(row[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcontent\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[0;32m      6\u001b[0m \u001b[38;5;66;03m# Display the DataFrame with the new summary column\u001b[39;00m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;28mprint\u001b[39m(df)\n",
      "Cell \u001b[1;32mIn[12], line 13\u001b[0m, in \u001b[0;36mgenerate_summary\u001b[1;34m(content)\u001b[0m\n\u001b[0;32m     10\u001b[0m prompt \u001b[38;5;241m=\u001b[39m ChatPromptTemplate\u001b[38;5;241m.\u001b[39mfrom_messages([(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msystem\u001b[39m\u001b[38;5;124m\"\u001b[39m, system), (\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhuman\u001b[39m\u001b[38;5;124m\"\u001b[39m, content)])\n\u001b[0;32m     12\u001b[0m chain \u001b[38;5;241m=\u001b[39m prompt \u001b[38;5;241m|\u001b[39m llm\n\u001b[1;32m---> 13\u001b[0m response \u001b[38;5;241m=\u001b[39m chain\u001b[38;5;241m.\u001b[39minvoke({})\n\u001b[0;32m     14\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m response\u001b[38;5;241m.\u001b[39mcontent\n",
      "File \u001b[1;32m~\\miniconda3\\Lib\\site-packages\\langchain_core\\runnables\\base.py:2873\u001b[0m, in \u001b[0;36mRunnableSequence.invoke\u001b[1;34m(self, input, config, **kwargs)\u001b[0m\n\u001b[0;32m   2871\u001b[0m             \u001b[38;5;28minput\u001b[39m \u001b[38;5;241m=\u001b[39m step\u001b[38;5;241m.\u001b[39minvoke(\u001b[38;5;28minput\u001b[39m, config, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m   2872\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m-> 2873\u001b[0m             \u001b[38;5;28minput\u001b[39m \u001b[38;5;241m=\u001b[39m step\u001b[38;5;241m.\u001b[39minvoke(\u001b[38;5;28minput\u001b[39m, config)\n\u001b[0;32m   2874\u001b[0m \u001b[38;5;66;03m# finish the root run\u001b[39;00m\n\u001b[0;32m   2875\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[1;32m~\\miniconda3\\Lib\\site-packages\\langchain_core\\language_models\\chat_models.py:265\u001b[0m, in \u001b[0;36mBaseChatModel.invoke\u001b[1;34m(self, input, config, stop, **kwargs)\u001b[0m\n\u001b[0;32m    254\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21minvoke\u001b[39m(\n\u001b[0;32m    255\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m    256\u001b[0m     \u001b[38;5;28minput\u001b[39m: LanguageModelInput,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    260\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[0;32m    261\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m BaseMessage:\n\u001b[0;32m    262\u001b[0m     config \u001b[38;5;241m=\u001b[39m ensure_config(config)\n\u001b[0;32m    263\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m cast(\n\u001b[0;32m    264\u001b[0m         ChatGeneration,\n\u001b[1;32m--> 265\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgenerate_prompt(\n\u001b[0;32m    266\u001b[0m             [\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_convert_input(\u001b[38;5;28minput\u001b[39m)],\n\u001b[0;32m    267\u001b[0m             stop\u001b[38;5;241m=\u001b[39mstop,\n\u001b[0;32m    268\u001b[0m             callbacks\u001b[38;5;241m=\u001b[39mconfig\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcallbacks\u001b[39m\u001b[38;5;124m\"\u001b[39m),\n\u001b[0;32m    269\u001b[0m             tags\u001b[38;5;241m=\u001b[39mconfig\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtags\u001b[39m\u001b[38;5;124m\"\u001b[39m),\n\u001b[0;32m    270\u001b[0m             metadata\u001b[38;5;241m=\u001b[39mconfig\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmetadata\u001b[39m\u001b[38;5;124m\"\u001b[39m),\n\u001b[0;32m    271\u001b[0m             run_name\u001b[38;5;241m=\u001b[39mconfig\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrun_name\u001b[39m\u001b[38;5;124m\"\u001b[39m),\n\u001b[0;32m    272\u001b[0m             run_id\u001b[38;5;241m=\u001b[39mconfig\u001b[38;5;241m.\u001b[39mpop(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrun_id\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m),\n\u001b[0;32m    273\u001b[0m             \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[0;32m    274\u001b[0m         )\u001b[38;5;241m.\u001b[39mgenerations[\u001b[38;5;241m0\u001b[39m][\u001b[38;5;241m0\u001b[39m],\n\u001b[0;32m    275\u001b[0m     )\u001b[38;5;241m.\u001b[39mmessage\n",
      "File \u001b[1;32m~\\miniconda3\\Lib\\site-packages\\langchain_core\\language_models\\chat_models.py:698\u001b[0m, in \u001b[0;36mBaseChatModel.generate_prompt\u001b[1;34m(self, prompts, stop, callbacks, **kwargs)\u001b[0m\n\u001b[0;32m    690\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mgenerate_prompt\u001b[39m(\n\u001b[0;32m    691\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m    692\u001b[0m     prompts: List[PromptValue],\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    695\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[0;32m    696\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m LLMResult:\n\u001b[0;32m    697\u001b[0m     prompt_messages \u001b[38;5;241m=\u001b[39m [p\u001b[38;5;241m.\u001b[39mto_messages() \u001b[38;5;28;01mfor\u001b[39;00m p \u001b[38;5;129;01min\u001b[39;00m prompts]\n\u001b[1;32m--> 698\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgenerate(prompt_messages, stop\u001b[38;5;241m=\u001b[39mstop, callbacks\u001b[38;5;241m=\u001b[39mcallbacks, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32m~\\miniconda3\\Lib\\site-packages\\langchain_core\\language_models\\chat_models.py:555\u001b[0m, in \u001b[0;36mBaseChatModel.generate\u001b[1;34m(self, messages, stop, callbacks, tags, metadata, run_name, run_id, **kwargs)\u001b[0m\n\u001b[0;32m    553\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m run_managers:\n\u001b[0;32m    554\u001b[0m             run_managers[i]\u001b[38;5;241m.\u001b[39mon_llm_error(e, response\u001b[38;5;241m=\u001b[39mLLMResult(generations\u001b[38;5;241m=\u001b[39m[]))\n\u001b[1;32m--> 555\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[0;32m    556\u001b[0m flattened_outputs \u001b[38;5;241m=\u001b[39m [\n\u001b[0;32m    557\u001b[0m     LLMResult(generations\u001b[38;5;241m=\u001b[39m[res\u001b[38;5;241m.\u001b[39mgenerations], llm_output\u001b[38;5;241m=\u001b[39mres\u001b[38;5;241m.\u001b[39mllm_output)  \u001b[38;5;66;03m# type: ignore[list-item]\u001b[39;00m\n\u001b[0;32m    558\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m results\n\u001b[0;32m    559\u001b[0m ]\n\u001b[0;32m    560\u001b[0m llm_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_combine_llm_outputs([res\u001b[38;5;241m.\u001b[39mllm_output \u001b[38;5;28;01mfor\u001b[39;00m res \u001b[38;5;129;01min\u001b[39;00m results])\n",
      "File \u001b[1;32m~\\miniconda3\\Lib\\site-packages\\langchain_core\\language_models\\chat_models.py:545\u001b[0m, in \u001b[0;36mBaseChatModel.generate\u001b[1;34m(self, messages, stop, callbacks, tags, metadata, run_name, run_id, **kwargs)\u001b[0m\n\u001b[0;32m    542\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(messages):\n\u001b[0;32m    543\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m    544\u001b[0m         results\u001b[38;5;241m.\u001b[39mappend(\n\u001b[1;32m--> 545\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_generate_with_cache(\n\u001b[0;32m    546\u001b[0m                 m,\n\u001b[0;32m    547\u001b[0m                 stop\u001b[38;5;241m=\u001b[39mstop,\n\u001b[0;32m    548\u001b[0m                 run_manager\u001b[38;5;241m=\u001b[39mrun_managers[i] \u001b[38;5;28;01mif\u001b[39;00m run_managers \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m    549\u001b[0m                 \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[0;32m    550\u001b[0m             )\n\u001b[0;32m    551\u001b[0m         )\n\u001b[0;32m    552\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mBaseException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m    553\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m run_managers:\n",
      "File \u001b[1;32m~\\miniconda3\\Lib\\site-packages\\langchain_core\\language_models\\chat_models.py:770\u001b[0m, in \u001b[0;36mBaseChatModel._generate_with_cache\u001b[1;34m(self, messages, stop, run_manager, **kwargs)\u001b[0m\n\u001b[0;32m    768\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    769\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m inspect\u001b[38;5;241m.\u001b[39msignature(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_generate)\u001b[38;5;241m.\u001b[39mparameters\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrun_manager\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m--> 770\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_generate(\n\u001b[0;32m    771\u001b[0m             messages, stop\u001b[38;5;241m=\u001b[39mstop, run_manager\u001b[38;5;241m=\u001b[39mrun_manager, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs\n\u001b[0;32m    772\u001b[0m         )\n\u001b[0;32m    773\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    774\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_generate(messages, stop\u001b[38;5;241m=\u001b[39mstop, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32m~\\miniconda3\\Lib\\site-packages\\langchain_groq\\chat_models.py:472\u001b[0m, in \u001b[0;36mChatGroq._generate\u001b[1;34m(self, messages, stop, run_manager, **kwargs)\u001b[0m\n\u001b[0;32m    467\u001b[0m message_dicts, params \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_create_message_dicts(messages, stop)\n\u001b[0;32m    468\u001b[0m params \u001b[38;5;241m=\u001b[39m {\n\u001b[0;32m    469\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mparams,\n\u001b[0;32m    470\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[0;32m    471\u001b[0m }\n\u001b[1;32m--> 472\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclient\u001b[38;5;241m.\u001b[39mcreate(messages\u001b[38;5;241m=\u001b[39mmessage_dicts, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mparams)\n\u001b[0;32m    473\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_create_chat_result(response)\n",
      "File \u001b[1;32m~\\miniconda3\\Lib\\site-packages\\groq\\resources\\chat\\completions.py:289\u001b[0m, in \u001b[0;36mCompletions.create\u001b[1;34m(self, messages, model, frequency_penalty, function_call, functions, logit_bias, logprobs, max_tokens, n, parallel_tool_calls, presence_penalty, response_format, seed, stop, stream, temperature, tool_choice, tools, top_logprobs, top_p, user, extra_headers, extra_query, extra_body, timeout)\u001b[0m\n\u001b[0;32m    148\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcreate\u001b[39m(\n\u001b[0;32m    149\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m    150\u001b[0m     \u001b[38;5;241m*\u001b[39m,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    177\u001b[0m     timeout: \u001b[38;5;28mfloat\u001b[39m \u001b[38;5;241m|\u001b[39m httpx\u001b[38;5;241m.\u001b[39mTimeout \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m|\u001b[39m NotGiven \u001b[38;5;241m=\u001b[39m NOT_GIVEN,\n\u001b[0;32m    178\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ChatCompletion \u001b[38;5;241m|\u001b[39m Stream[ChatCompletionChunk]:\n\u001b[0;32m    179\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    180\u001b[0m \u001b[38;5;124;03m    Creates a model response for the given chat conversation.\u001b[39;00m\n\u001b[0;32m    181\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    287\u001b[0m \u001b[38;5;124;03m      timeout: Override the client-level default timeout for this request, in seconds\u001b[39;00m\n\u001b[0;32m    288\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 289\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_post(\n\u001b[0;32m    290\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/openai/v1/chat/completions\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m    291\u001b[0m         body\u001b[38;5;241m=\u001b[39mmaybe_transform(\n\u001b[0;32m    292\u001b[0m             {\n\u001b[0;32m    293\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmessages\u001b[39m\u001b[38;5;124m\"\u001b[39m: messages,\n\u001b[0;32m    294\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmodel\u001b[39m\u001b[38;5;124m\"\u001b[39m: model,\n\u001b[0;32m    295\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfrequency_penalty\u001b[39m\u001b[38;5;124m\"\u001b[39m: frequency_penalty,\n\u001b[0;32m    296\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfunction_call\u001b[39m\u001b[38;5;124m\"\u001b[39m: function_call,\n\u001b[0;32m    297\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfunctions\u001b[39m\u001b[38;5;124m\"\u001b[39m: functions,\n\u001b[0;32m    298\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlogit_bias\u001b[39m\u001b[38;5;124m\"\u001b[39m: logit_bias,\n\u001b[0;32m    299\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mlogprobs\u001b[39m\u001b[38;5;124m\"\u001b[39m: logprobs,\n\u001b[0;32m    300\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmax_tokens\u001b[39m\u001b[38;5;124m\"\u001b[39m: max_tokens,\n\u001b[0;32m    301\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mn\u001b[39m\u001b[38;5;124m\"\u001b[39m: n,\n\u001b[0;32m    302\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mparallel_tool_calls\u001b[39m\u001b[38;5;124m\"\u001b[39m: parallel_tool_calls,\n\u001b[0;32m    303\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpresence_penalty\u001b[39m\u001b[38;5;124m\"\u001b[39m: presence_penalty,\n\u001b[0;32m    304\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresponse_format\u001b[39m\u001b[38;5;124m\"\u001b[39m: response_format,\n\u001b[0;32m    305\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mseed\u001b[39m\u001b[38;5;124m\"\u001b[39m: seed,\n\u001b[0;32m    306\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstop\u001b[39m\u001b[38;5;124m\"\u001b[39m: stop,\n\u001b[0;32m    307\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstream\u001b[39m\u001b[38;5;124m\"\u001b[39m: stream,\n\u001b[0;32m    308\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtemperature\u001b[39m\u001b[38;5;124m\"\u001b[39m: temperature,\n\u001b[0;32m    309\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtool_choice\u001b[39m\u001b[38;5;124m\"\u001b[39m: tool_choice,\n\u001b[0;32m    310\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtools\u001b[39m\u001b[38;5;124m\"\u001b[39m: tools,\n\u001b[0;32m    311\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtop_logprobs\u001b[39m\u001b[38;5;124m\"\u001b[39m: top_logprobs,\n\u001b[0;32m    312\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtop_p\u001b[39m\u001b[38;5;124m\"\u001b[39m: top_p,\n\u001b[0;32m    313\u001b[0m                 \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124muser\u001b[39m\u001b[38;5;124m\"\u001b[39m: user,\n\u001b[0;32m    314\u001b[0m             },\n\u001b[0;32m    315\u001b[0m             completion_create_params\u001b[38;5;241m.\u001b[39mCompletionCreateParams,\n\u001b[0;32m    316\u001b[0m         ),\n\u001b[0;32m    317\u001b[0m         options\u001b[38;5;241m=\u001b[39mmake_request_options(\n\u001b[0;32m    318\u001b[0m             extra_headers\u001b[38;5;241m=\u001b[39mextra_headers, extra_query\u001b[38;5;241m=\u001b[39mextra_query, extra_body\u001b[38;5;241m=\u001b[39mextra_body, timeout\u001b[38;5;241m=\u001b[39mtimeout\n\u001b[0;32m    319\u001b[0m         ),\n\u001b[0;32m    320\u001b[0m         cast_to\u001b[38;5;241m=\u001b[39mChatCompletion,\n\u001b[0;32m    321\u001b[0m         stream\u001b[38;5;241m=\u001b[39mstream \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[0;32m    322\u001b[0m         stream_cls\u001b[38;5;241m=\u001b[39mStream[ChatCompletionChunk],\n\u001b[0;32m    323\u001b[0m     )\n",
      "File \u001b[1;32m~\\miniconda3\\Lib\\site-packages\\groq\\_base_client.py:1225\u001b[0m, in \u001b[0;36mSyncAPIClient.post\u001b[1;34m(self, path, cast_to, body, options, files, stream, stream_cls)\u001b[0m\n\u001b[0;32m   1211\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mpost\u001b[39m(\n\u001b[0;32m   1212\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m   1213\u001b[0m     path: \u001b[38;5;28mstr\u001b[39m,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1220\u001b[0m     stream_cls: \u001b[38;5;28mtype\u001b[39m[_StreamT] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m   1221\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ResponseT \u001b[38;5;241m|\u001b[39m _StreamT:\n\u001b[0;32m   1222\u001b[0m     opts \u001b[38;5;241m=\u001b[39m FinalRequestOptions\u001b[38;5;241m.\u001b[39mconstruct(\n\u001b[0;32m   1223\u001b[0m         method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpost\u001b[39m\u001b[38;5;124m\"\u001b[39m, url\u001b[38;5;241m=\u001b[39mpath, json_data\u001b[38;5;241m=\u001b[39mbody, files\u001b[38;5;241m=\u001b[39mto_httpx_files(files), \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39moptions\n\u001b[0;32m   1224\u001b[0m     )\n\u001b[1;32m-> 1225\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m cast(ResponseT, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrequest(cast_to, opts, stream\u001b[38;5;241m=\u001b[39mstream, stream_cls\u001b[38;5;241m=\u001b[39mstream_cls))\n",
      "File \u001b[1;32m~\\miniconda3\\Lib\\site-packages\\groq\\_base_client.py:920\u001b[0m, in \u001b[0;36mSyncAPIClient.request\u001b[1;34m(self, cast_to, options, remaining_retries, stream, stream_cls)\u001b[0m\n\u001b[0;32m    911\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrequest\u001b[39m(\n\u001b[0;32m    912\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[0;32m    913\u001b[0m     cast_to: Type[ResponseT],\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    918\u001b[0m     stream_cls: \u001b[38;5;28mtype\u001b[39m[_StreamT] \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[0;32m    919\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m ResponseT \u001b[38;5;241m|\u001b[39m _StreamT:\n\u001b[1;32m--> 920\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_request(\n\u001b[0;32m    921\u001b[0m         cast_to\u001b[38;5;241m=\u001b[39mcast_to,\n\u001b[0;32m    922\u001b[0m         options\u001b[38;5;241m=\u001b[39moptions,\n\u001b[0;32m    923\u001b[0m         stream\u001b[38;5;241m=\u001b[39mstream,\n\u001b[0;32m    924\u001b[0m         stream_cls\u001b[38;5;241m=\u001b[39mstream_cls,\n\u001b[0;32m    925\u001b[0m         remaining_retries\u001b[38;5;241m=\u001b[39mremaining_retries,\n\u001b[0;32m    926\u001b[0m     )\n",
      "File \u001b[1;32m~\\miniconda3\\Lib\\site-packages\\groq\\_base_client.py:1018\u001b[0m, in \u001b[0;36mSyncAPIClient._request\u001b[1;34m(self, cast_to, options, remaining_retries, stream, stream_cls)\u001b[0m\n\u001b[0;32m   1015\u001b[0m         err\u001b[38;5;241m.\u001b[39mresponse\u001b[38;5;241m.\u001b[39mread()\n\u001b[0;32m   1017\u001b[0m     log\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mRe-raising status error\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m-> 1018\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_make_status_error_from_response(err\u001b[38;5;241m.\u001b[39mresponse) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m   1020\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_process_response(\n\u001b[0;32m   1021\u001b[0m     cast_to\u001b[38;5;241m=\u001b[39mcast_to,\n\u001b[0;32m   1022\u001b[0m     options\u001b[38;5;241m=\u001b[39moptions,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1025\u001b[0m     stream_cls\u001b[38;5;241m=\u001b[39mstream_cls,\n\u001b[0;32m   1026\u001b[0m )\n",
      "\u001b[1;31mBadRequestError\u001b[0m: Error code: 400 - {'error': {'message': 'Please reduce the length of the messages or completion.', 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "# Add the summary column using a for loop\n",
    "for index, row in df.iterrows():\n",
    "    df.at[index, 'summary'] = generate_summary(row['content'])\n",
    "\n",
    "\n",
    "# Display the DataFrame with the new summary column\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Molecular Docking As A Potential Approach In Repurposing Drugs Against Covid19: A Systematic Review And Novel Pharmacophore Models\n",
      "## Abstract\n",
      "## Introduction\n",
      "## Methods Workfow Search Strategy\n",
      "## Data Selection And Extraction\n",
      "## Pharmacophore Analysis\n",
      "## Data Description\n",
      "## Results And Discussion Main Protease\n",
      "## Antivirals\n",
      "## Tetracyclines\n",
      "## Pharmacophore Analysis Of The Main Protease\n",
      "## RnaDependent Rna Polymerase\n",
      "## Drug Afecting RnaDependent Rna Polymerase\n",
      "## Pharmacophore Analysis Of The RnaDependent Rna Polymerase\n",
      "## Spike Protein\n",
      "## Drug Afecting Spike Protein\n",
      "## Drug Afecting The Ace2/Spike Complex\n",
      "## Pharmacophore Analysis Of The Spike Protein\n",
      "## Conclusion\n",
      "## Declarations\n",
      "## References\n"
     ]
    },
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'ace_tools'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[18], line 33\u001b[0m\n\u001b[0;32m     30\u001b[0m \u001b[38;5;66;03m# Apply the function to the dataframe\u001b[39;00m\n\u001b[0;32m     31\u001b[0m df_with_citations_and_summaries \u001b[38;5;241m=\u001b[39m count_citations_and_generate_summary(df)\n\u001b[1;32m---> 33\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mace_tools\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mtools\u001b[39;00m; tools\u001b[38;5;241m.\u001b[39mdisplay_dataframe_to_user(name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTitle, Word Count, Citation Count, and Summary\u001b[39m\u001b[38;5;124m\"\u001b[39m, dataframe\u001b[38;5;241m=\u001b[39mdf_with_citations_and_summaries)\n\u001b[0;32m     35\u001b[0m \u001b[38;5;66;03m# Display the modified dataframe\u001b[39;00m\n\u001b[0;32m     36\u001b[0m df_with_citations_and_summaries\u001b[38;5;241m.\u001b[39mhead()\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'ace_tools'"
     ]
    }
   ],
   "source": [
    "import util\n",
    "import importlib\n",
    "importlib.reload(util)\n",
    "\n",
    "def count_citations_and_generate_summary(df):\n",
    "    # Function to check if the title contains reference-related words and count DOI occurrences\n",
    "    reference_keywords = ['References', 'Citation', 'Citations', 'Bibliography', 'References', 'Reference']\n",
    "    \n",
    "    # Check if the title contains any reference-related words\n",
    "    def is_reference_title(title):\n",
    "        return any(keyword.lower() in title.lower() for keyword in reference_keywords)\n",
    "    \n",
    "    # Count DOI occurrences in the content\n",
    "    def count_doi(content):\n",
    "        return content.lower().count('doi')\n",
    "    \n",
    "    # Create new columns for citation counts and summaries\n",
    "    df['summary'] = \"\"\n",
    "    \n",
    "    for index, row in df.iterrows():\n",
    "        print(row['title'])\n",
    "        if is_reference_title(row['title']):\n",
    "            citations = count_doi(row['content'])\n",
    "            df.at[index, 'summary'] = \"Number of citaiton: {citations}\"\n",
    "        else:\n",
    "            df.at[index, 'summary'] = generate_summary(row['content'])\n",
    "    \n",
    "    return df\n",
    "\n",
    "# Apply the function to the dataframe\n",
    "df_with_citations_and_summaries = count_citations_and_generate_summary(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "**Molecular Docking as a Potential Approach in Repurposing Drugs Against COVID-19: A Systematic Review and Novel Pharmacophore Models**\n",
       "=====================================================\n",
       "\n",
       "### Abstract\n",
       "\n",
       "This systematic review explores the potential of molecular docking as an approach to repurpose drugs against COVID-19. The study aims to identify novel pharmacophore models that can aid in the discovery of new therapeutic agents.\n",
       "\n",
       "### Introduction\n",
       "\n",
       "The COVID-19 pandemic has highlighted the need for innovative approaches to drug discovery and repurposing. Molecular docking is a promising technique that can be used to identify potential therapeutic agents by predicting their binding affinity to target proteins.\n",
       "\n",
       "### Methods\n",
       "\n",
       "A systematic review was conducted to identify relevant studies on molecular docking and COVID-19. The search strategy included major databases and keywords related to molecular docking, COVID-19, and drug repurposing.\n",
       "\n",
       "### Results\n",
       "\n",
       "The review identified several studies that have used molecular docking to identify potential therapeutic agents against COVID-19. The results showed that molecular docking can be a valuable tool in the discovery of new therapeutic agents.\n",
       "\n",
       "### Novel Pharmacophore Models\n",
       "\n",
       "The study also developed novel pharmacophore models that can be used to identify potential therapeutic agents against COVID-19. The models were developed using a combination of molecular docking and machine learning algorithms.\n",
       "\n",
       "### Discussion\n",
       "\n",
       "The results of this study demonstrate the potential of molecular docking as an approach to repurpose drugs against COVID-19. The novel pharmacophore models developed in this study can be used to identify potential therapeutic agents and aid in the discovery of new therapeutic agents.\n",
       "\n",
       "### References\n",
       "\n",
       "[Insert references]\n",
       "\n",
       "### Figure\n",
       "\n",
       "![0_image_0.png](0_image_0.png)\n",
       "\n",
       "Note: The figure is not included in this summary as it is an image. **Summary of Recent Literature on Repurposing FDA-Approved Drugs for COVID-19 Treatment**\n",
       "======================================================\n",
       "\n",
       "### Purpose\n",
       "The purpose of this review is to summarize the recent literature on FDA-approved drugs that have been repurposed as potential candidates against COVID-19. The study also includes a quality pharmacophore analysis of frequently studied targets, including the main protease, RNA-dependent RNA polymerase, and spike protein.\n",
       "\n",
       "### Recent Findings\n",
       "The COVID-19 pandemic has highlighted the need for effective therapeutic agents. Despite ongoing research, no absolute therapeutic agent has been developed. Drug repurposing strategies offer a viable approach to develop novel therapeutic interventions.\n",
       "\n",
       "### Drug Repurposing\n",
       "Drug repurposing involves exploring previously approved drugs with known safety and pharmacokinetic profiles for new effects, reducing costs, time, and predicting potential side effects and drug interactions. The COVID-19 viral machinery shares similarities with other viruses, making antiviral agents widely repurposed in the search for curative candidates.\n",
       "\n",
       "### Pharmacophore Analysis\n",
       "The main protease pharmacophoric study revealed multiple features, providing a starting point for future research.\n",
       "\n",
       "**Keywords**\n",
       "- COVID-19\n",
       "- Drug repurposing\n",
       "- Pharmacophore analysis\n",
       "- Main protease\n",
       "- Spike proteins\n",
       "- RNA-dependent RNA polymerase **Summary**\n",
       "===============\n",
       "\n",
       "**Introduction**\n",
       "---------------\n",
       "\n",
       "In late 2019, a novel beta coronavirus was identified as the cause of severe pneumonia cases in Wuhan, China. The World Health Organization (WHO) labeled the disease as COVID-19 and the virus as SARS-CoV-2.\n",
       "\n",
       "**Global Pandemic**\n",
       "-----------------\n",
       "\n",
       "By March 2020, the WHO announced that COVID-19 had become a global pandemic, affecting the entire world. Due to its airborne infectivity, social distancing measures were imposed by international authorities.\n",
       "\n",
       "**Drug Repurposing**\n",
       "-----------------\n",
       "\n",
       "In response to the sudden outbreak, global research efforts focused on investigating potential therapeutic approaches. Drug repurposing strategies were implemented as an immediate drug discovery pipeline. Clinically useful drugs, such as aspirin and sildenafil, were successfully repurposed for cardiovascular diseases and erectile dysfunction, respectively.\n",
       "\n",
       "**In Silico Methods**\n",
       "-------------------\n",
       "\n",
       "Molecular docking and pharmacophore modeling are popular in silico methods used in drug discovery research. Molecular docking illustrates interactions between ligands and biological targets, while pharmacophore modeling identifies molecular features that enable biologically active ligands to exert a pharmacological effect.\n",
       "\n",
       "**Research Summary**\n",
       "-------------------\n",
       "\n",
       "This article reviews and summarizes research on FDA-approved drugs repurposed against SARS-CoV-2 virus. **Methods Workflow Search Strategy**\n",
       "===============================\n",
       "\n",
       "### Background\n",
       "During the COVID-19 pandemic, the scientific community published a vast number of studies on repurposing drugs for COVID-19. These studies are scattered across the internet, making it challenging to find a comprehensive database that includes all of them, aside from Google Scholar.\n",
       "\n",
       "### Search Strategy\n",
       "To achieve an efficient search strategy, the following combinations of keywords were applied during the research in the Google Scholar database:\n",
       "\n",
       "* `(COVID19 OR SARS-CoV-2)` to capture studies related to COVID-19\n",
       "* `(Drug repurposing OR Drug repositioning OR Drug re-profiling OR Drug rediscovery)` to focus on studies about repurposing drugs\n",
       "* `(Docking AND Molecular dynamic)` to narrow down the search to studies that involve docking and molecular dynamics\n",
       "\n",
       "### Search Parameters\n",
       "The search was conducted up to December 30, 2020, to ensure that all relevant studies were included. **Data Selection And Extraction**\n",
       "=============================\n",
       "\n",
       "### Methodology\n",
       "\n",
       "* Titles and abstracts were reviewed for eligibility using specific inclusion and exclusion criteria (Table 1)\n",
       "* Selected papers were further reviewed in a second stage, while articles that met the exclusion criteria were eliminated\n",
       "\n",
       "### Search Results\n",
       "-----------------\n",
       "\n",
       "* Initial search yielded 405 articles\n",
       "* After applying inclusion and exclusion criteria, 92 research articles were selected\n",
       "* The selection process is presented in a flow diagram (Fig. 1)\n",
       "\n",
       "### Classification\n",
       "-----------------\n",
       "\n",
       "* The 92 selected articles were classified according to their target:\n",
       "\t+ 73 articles focused on the main protease\n",
       "\t+ 11 articles focused on the spike protein\n",
       "\t+ 8 articles focused on the replicase complex **Pharmacophore Analysis**\n",
       "======================\n",
       "\n",
       "### Introduction\n",
       "\n",
       "To perform pharmacophore analysis, it was essential to ensure that all analyzed drugs bind to the same site on the target. Therefore, drugs were classified within the same target according to the binding site.\n",
       "\n",
       "### Binding Site Definition\n",
       "\n",
       "Various approaches are used in literature to define the binding site or docking site. In this study, one binding site was obtained for each target, except for the main protease, which yielded two binding sites.\n",
       "\n",
       "### Pharmacophore Generation\n",
       "\n",
       "The Phase tool from Schrdinger was used to generate the pharmacophore hypothesis. The quality of the pharmacophore was a major concern, and the number of features and Hyposcore were used as measures of quality.\n",
       "\n",
       "### Quality Criteria\n",
       "\n",
       "The following criteria were implemented to ensure the quality of the pharmacophore:\n",
       "\n",
       "* Maximum of 7 features\n",
       "* Minimum of 6 features\n",
       "* Minimum coverage of 50% of the compounds at the question **Summary of In Silico Work on COVID-19 Drug Discovery**\n",
       "======================================================\n",
       "\n",
       "### Background\n",
       "\n",
       "In silico work on COVID-19 drug discovery began before the release of the PDB structure of SARS-CoV-2 targets. Homology modeling was used to generate protein structures, focusing on the spike protein, RNA-dependent RNA polymerase (RdRp), main protease, helicase, and papain-like protease. This summary reviews the most studied targets: main protease, spike protein, and RdRp.\n",
       "\n",
       "### Methods\n",
       "\n",
       "Papers followed a similar protocol, using either docking alone or docking and molecular dynamics simulation. The most used software for molecular docking was Autodock vina, followed by Glide and Autodock 4. Autodock vina was the most used due to its ease of use, open-source feature, and accuracy.\n",
       "\n",
       "### Molecular Dynamics Simulation\n",
       "\n",
       "Molecular dynamics (MD) simulation takes the ligand-protein complex as a dynamic module, searching for the most stable conformation. Only 48 out of 89 studies confirmed docking results with MD.\n",
       "\n",
       "### Inclusion and Exclusion Criteria\n",
       "\n",
       "The following criteria were used to select articles:\n",
       "\n",
       "| Parameter | Inclusion Criteria | Exclusion Criteria |\n",
       "| --- | --- | --- |\n",
       "| Type of publication | Original article | Preprint, review article, letters, conference abstracts |\n",
       "| Kind of study | Studies that evaluated antiviral activities against COVID-19 | Studies that evaluated antiviral activities against other viruses |\n",
       "| Method of study | Docking, molecular dynamic | Homology modeling |\n",
       "| Type of molecule | Drugs | Phytochemicals, non-drug molecules |\n",
       "| Language | English | Language other than English |\n",
       "\n",
       "### Flow Diagram\n",
       "\n",
       "The flow diagram (Fig. 1) illustrates the screening process of articles included in the study.\n",
       "\n",
       "### Software Used\n",
       "\n",
       "Autodock vina was the most used software for molecular docking, followed by Glide and Autodock 4. **Results and Discussion: Main Protease**\n",
       "================================================\n",
       "\n",
       "### Main Protease Target\n",
       "\n",
       "The main protease of SARS-CoV-2 has been identified as a well-positioned drug target due to its unique cleavage activity, which is not mimicked by human proteases [13].\n",
       "\n",
       "### Sequence Conservation\n",
       "\n",
       "The main protease of SARS-CoV-2 shows a high sequence identity of 96% with other SARS-CoV viruses [14], encouraging researchers to study inhibitors and repurpose existing drugs against the SARS-CoV-2 main protease.\n",
       "\n",
       "### Crystal Structure\n",
       "\n",
       "The crystal structure of the main protease was first released by Yang's group (PDB ID: 6LU7) [15].\n",
       "\n",
       "### Drug Classification\n",
       "\n",
       "Drugs were classified according to the Anatomical Therapeutic Chemical (ATC) classification system (Table 2).\n",
       "\n",
       "### Results\n",
       "\n",
       "* Anti-infective agents were the most frequent class, with antivirals and tetracycline being the predominant subclasses.\n",
       "* Antineoplastic and immunomodulating agents were the next most frequent classes (Table 2). **Antivirals**\n",
       "================\n",
       "\n",
       "### Introduction\n",
       "\n",
       "The study investigates the potential of 29 antiviral drugs that bind with the viral main protease to also inhibit SARS-COV-2 protease. Of these, 16 drugs are already protease inhibitors, including indinavir, ritonavir, lopinavir, and others.\n",
       "\n",
       "### Methodology\n",
       "\n",
       "The study used docking software to analyze the binding of these drugs with the SARS-COV-2 protease. Indinavir, ritonavir, and lopinavir were tested multiple times with different software, providing evidence for their potential as COVID-19 drugs.\n",
       "\n",
       "### Results\n",
       "\n",
       "* 16 drugs are already protease inhibitors, including indinavir, ritonavir, lopinavir, and others.\n",
       "* Indinavir, ritonavir, and lopinavir were tested multiple times with different software, providing evidence for their potential as COVID-19 drugs.\n",
       "* Asunaprevir, ciluprevir, and pleconaril work on \"Genome polyprotein\" and could be considered for further analysis.\n",
       "* The remaining 10 drugs have diverse targets and failed pharmacophore analysis, and were excluded from further analysis.\n",
       "\n",
       "### Discussion\n",
       "\n",
       "The study provides evidence for the potential of certain antiviral drugs to inhibit SARS-COV-2 protease, and suggests that further analysis of asunaprevir, ciluprevir, and pleconaril may be warranted. **Tetracyclines**\n",
       "================\n",
       "\n",
       "### Overview\n",
       "\n",
       "Five tetracyclines have been reported to exhibit inhibitory activity against the main protease: oxytetracycline, lymecycline, tetracycline, minocycline, and doxycycline.\n",
       "\n",
       "### Structure and Mechanism of Action\n",
       "\n",
       "Tetracyclines are broad-spectrum antibiotics with a general structure consisting of four fused rings. They work by inhibiting protein synthesis, specifically by preventing the binding of aminoacyl-tRNA to the mRNA translation complex. Here is a summarized version of the text in a structured markdown format:\n",
       "\n",
       "**Pharmacophore Analysis of the Main Protease**\n",
       "=============================================\n",
       "\n",
       "### Introduction\n",
       "\n",
       "The main protease has been studied extensively, with most studies focusing on the 6LU7 structure. This study aims to analyze the pharmacophore of the main protease and identify potential inhibitors.\n",
       "\n",
       "### Results\n",
       "\n",
       "* The pharmacophore of the main protease was analyzed using a dataset of 10,000 compounds.\n",
       "* The results showed that tetracycline and N3 inhibitor were the most similar pair of hypotheses, with 4 matched features.\n",
       "* The pharmacophores of tetracycline and N3 inhibitor were found to align with at least 3 matches, except for N3 inhibitor and tetracycline, which aligned with 4 matched features.\n",
       "* The overlapping features of N3 and tetracycline were identified as 2 hydrogen bond donors, 1 hydrogen bond acceptor, and 1 aromatic ring (Fig. 3a).\n",
       "\n",
       "### Discussion\n",
       "\n",
       "* The results suggest that tetracycline and N3 inhibitor are potential inhibitors of the main protease.\n",
       "* The pharmacophore analysis provides a framework for identifying potential inhibitors of the main protease.\n",
       "* Further studies are needed to validate the results and identify additional inhibitors.\n",
       "\n",
       "### References\n",
       "\n",
       "* [17, 18, 24, 25]\n",
       "* [34, 35]\n",
       "* [19, 46]\n",
       "* [31]\n",
       "* [16]\n",
       "* [20]\n",
       "* [21]\n",
       "* [22]\n",
       "* [23]\n",
       "* [26]\n",
       "* [27]\n",
       "* [28]\n",
       "* [29]\n",
       "* [30]\n",
       "* [32]\n",
       "* [33]\n",
       "* [38]\n",
       "* [39]\n",
       "* [40]\n",
       "* [41]\n",
       "* [42]\n",
       "* [43]\n",
       "* [44]\n",
       "* [45]\n",
       "* [46]\n",
       "* [47]\n",
       "* [48]\n",
       "* [49]\n",
       "* [50]\n",
       "* [51]\n",
       "* [52]\n",
       "* [53]\n",
       "* [54]\n",
       "* [55]\n",
       "* [56]\n",
       "* [57]\n",
       "* [58]\n",
       "* [59]\n",
       "* [60]\n",
       "* [61]\n",
       "* [62]\n",
       "* [63]\n",
       "* [64]\n",
       "* [65]\n",
       "* [66]\n",
       "* [67]\n",
       "* [68]\n",
       "* [69]\n",
       "* [70]\n",
       "* [71]\n",
       "* [72]\n",
       "* [73]\n",
       "* [74]\n",
       "* [75]\n",
       "* [76]\n",
       "* [77]\n",
       "* [78]\n",
       "* [79]\n",
       "* [80]\n",
       "* [81]\n",
       "* [82]\n",
       "* [83]\n",
       "* [84]\n",
       "* [85]\n",
       "* [86]\n",
       "* [87]\n",
       "* [88]\n",
       "\n",
       "Note: The references are listed at the end of the summary. **RNA-Dependent RNA Polymerase (RdRp)**\n",
       "===============================\n",
       "\n",
       "### Overview\n",
       "\n",
       "RNA-dependent RNA polymerase (RdRp) is a replicase that synthesizes a complementary RNA strand using RNA as a template. The SARS-CoV-2 RdRp is composed of 932 amino acids, divided into the N-terminal and polymerase domains.\n",
       "\n",
       "### Structure\n",
       "\n",
       "The RdRp protein is structured into three subunits: the finger, palm, and thumb. The active site is located in the middle of the substrate domain, where RNA synthesis occurs.\n",
       "\n",
       "### Function\n",
       "\n",
       "The RdRp enzyme synthesizes RNA by accessing an RNA template from the template input channel and nucleoside triphosphate (NTP) from the NTP input channel.\n",
       "\n",
       "### Target for Drug Development\n",
       "\n",
       "RdRp has been proposed as an important target for developing drugs against coronavirus. **Drug Afecting Rna-Dependent Rna Polymerase**\n",
       "===============================\n",
       "\n",
       "### Introduction\n",
       "\n",
       "A review of literature revealed several papers discussing Nsp12, a key enzyme in the SARS-CoV-2 replication process. The study focused on antiviral agents that target the polymerase, as well as antineoplastic agents with potential inhibitory effects.\n",
       "\n",
       "### Antiviral Agents\n",
       "\n",
       "Five antiviral agents were identified as targeting the polymerase: remdesivir, tipranavir, tegobuvir, simeprevir, and flibuvir. These agents have diverse mechanisms of action and viral targets.\n",
       "\n",
       "### Remdesivir\n",
       "\n",
       "Remdesivir is a nucleoside analog that has been extensively studied in vitro and in silico. It is incorporated into the growing viral RNA, terminating RNA replication. Remdesivir has high affinity for the active site of RdRp, competing with natural substrates and getting incorporated into the RNA chain. It has been investigated in multiple COVID-19 clinical trials and has received FDA Emergency Use Authorization for COVID-19 treatment.\n",
       "\n",
       "### Other Reported Drugs\n",
       "\n",
       "While other reported drugs have no similarity in their original clinical use, in silico docking data suggests a definitive interaction with the Nsp12 active site, making them promising inhibitors.\n",
       "\n",
       "### Pharmacophore Analysis\n",
       "\n",
       "A pharmacophore analysis was performed based on drugs and the common binding site, despite the limited available data. **Pharmacophore Analysis of RNA-Dependent RNA Polymerase**\n",
       "======================================================\n",
       "\n",
       "### Introduction\n",
       "\n",
       "A pharmacophore analysis was conducted on eight FDA-approved drugs of miscellaneous pharmacological classes to identify potential targets for the RNA-dependent RNA polymerase enzyme.\n",
       "\n",
       "### Results\n",
       "\n",
       "The analysis revealed a six-feature pharmacophore, with the following features:\n",
       "* Hydrogen-bond donor: 1\n",
       "* Hydrogen-bond acceptor: 3\n",
       "* Aromatic ring: 2\n",
       "\n",
       "The most frequent target observed was the PDB ID 6M71.\n",
       "\n",
       "### Comparison with Drug-Based Analysis\n",
       "\n",
       "The results of the pharmacophore analysis were compared with those of a drug-based analysis. The two analyses disagreed on the presence of hydrogen-bond donors and acceptors, but agreed on the presence of aromatic rings.\n",
       "\n",
       "### Pharmacophore Features\n",
       "\n",
       "The pharmacophore analysis identified seven features, including:\n",
       "* 3 hydrogen-bond donors\n",
       "* 1 hydrogen-bond acceptor\n",
       "* 2 aromatic rings\n",
       "* 1 negative ionic feature (unique to the binding-site based hypothesis study)\n",
       "\n",
       "### Figure 3\n",
       "\n",
       "![8_image_0.png](8_image_0.png)\n",
       "\n",
       "**Figure 3: Pharmacophore Analysis for Targeting SARS-CoV-2 Enzymes**\n",
       "\n",
       "a. Main protease\n",
       "b. RNA-dependent RNA polymerase **Spike Protein**\n",
       "================\n",
       "\n",
       "### Overview\n",
       "\n",
       "The spike protein is a class fusion glycoprotein located on the surface of SARS-CoV-2. It plays a crucial role in viral infection by recognizing the host angiotensin-converting enzyme-2 receptor (ACE2).\n",
       "\n",
       "### Structure\n",
       "\n",
       "The spike protein consists of two subunits, S1 and S2. The S1 subunit contains a receptor-binding domain (RBD) that binds and recognizes the ACE2 receptor, while the S2 subunit mediates viral cell membrane fusion.\n",
       "\n",
       "### Function\n",
       "\n",
       "The spike protein uses the ACE-2 receptor for entry to the host cell. It is coated with polysaccharide molecules to evade the surveillance of the host immune system during entry. Targeting the spike protein can potentially decrease the chance of infection by preventing viral invasion to host cells.\n",
       "\n",
       "### Repurposed Drugs\n",
       "\n",
       "Research has been conducted to identify the role of repurposed drugs on spike proteins and ACE2. Table 4 lists drugs that target RNA-dependent RNA polymerase, while Table 5 classifies studies that interrupt the COVID-19 mechanism of entry into three classes:\n",
       "\n",
       "* Drugs that bind to the spike protein and cannot recognize ACE2\n",
       "* Drugs that bind to ACE2 and compete with the spike protein on the ACE2 binding site\n",
       "* Drugs that bind to the complex of spike and ACE2 and prevent the transition to the active state\n",
       "\n",
       "### Classification\n",
       "\n",
       "According to the ATC classification system, most of the drugs affecting the spike protein, ACE2, and the complex fall in the anti-infective category.\n",
       "\n",
       "### References\n",
       "\n",
       "[92-97, 98-100, 102, 106, 88] **Drug Affecting Spike Protein**\n",
       "=============================\n",
       "\n",
       "### Overview\n",
       "\n",
       "Thirty-four drugs have been reported to target the spike protein, with 6 of them being antivirals and 3 of them being nucleoside analogs that originally target DNA synthesis in viruses.\n",
       "\n",
       "### List of Drugs\n",
       "\n",
       "| PDB ID | Drug | References |\n",
       "| --- | --- | --- |\n",
       "| 6M17 | Ivermectin | [16] |\n",
       "| 6LZG | Hydroxychloroquine, Chloroquine | [101] |\n",
       "| 6VW1 | Troxerutin | [102] |\n",
       "| 6VSB | Dolutegravir | [102, 103] |\n",
       "| 6M17 | Cangrelor | [102, 103, 104] |\n",
       "| 6M17 | Kanamycin | [104] |\n",
       "| 6VSB | Nelfnavir | [104, 105] |\n",
       "| 2AJF | Lopinavir | [106] |\n",
       "| 6LZG | Hydroxychloroquine | [101] |\n",
       "| 6VSB | Daclatasvir | [42, 102] |\n",
       "| 6M0J | Polymyxin B | [88] |\n",
       "| 6M17 | Pralatrexate | [107] |\n",
       "| 6VXX | Imipenem | [107, 108] |\n",
       "| 6VSB | Streptomycin | [53] |\n",
       "\n",
       "### Notes\n",
       "\n",
       "* Ivermectin is reported twice in the spike docking analysis, but its frequency of appearance does not indicate a potential for being a lead compound.\n",
       "* Some in silico trials targeted ivermectin from the beginning, rather than as the end result of an exhaustive virtual screening. **Drug Afecting The ACE2/Spike Complex**\n",
       "=====================================\n",
       "\n",
       "### Overview\n",
       "\n",
       "Several studies have investigated the binding of drugs to the ACE2/spike complex, aiming to disrupt the interaction of ACE2 with RBD after spike binding has already occurred. Eight drugs have been reported to target this complex, including:\n",
       "\n",
       "* Naltrexone\n",
       "* Ivermectin\n",
       "* Lopinavir\n",
       "* Chloroquine\n",
       "* Polymyxin B\n",
       "* Colistin\n",
       "* Daptomycin\n",
       "* Oritavancin\n",
       "\n",
       "### Characteristics of Targeted Drugs\n",
       "\n",
       "* Molecular weight above 500, with most above 1000\n",
       "* Pharmacophore analysis failed due to complexities and variability\n",
       "\n",
       "### Studies Targeting ACE2 Receptor Only\n",
       "\n",
       "Two studies have investigated compounds that prevent the recognition of spike to ACE2. Ten drugs were reported to bind well to ACE2, including:\n",
       "\n",
       "* Lividomycin\n",
       "* Burixafor\n",
       "* Quisinostat\n",
       "* Fuprofylline\n",
       "* Pemetrexed\n",
       "* Spirofylline\n",
       "* Edotecarin\n",
       "* Diniprofylline\n",
       "* Hydroxychloroquine\n",
       "* Chloroquine\n",
       "\n",
       "Note: These compounds may bind to multiple targets, including ACE2. **Pharmacophore Analysis of the Spike Protein**\n",
       "=============================================\n",
       "\n",
       "### Background\n",
       "\n",
       "Pharmacophore analysis was conducted on the spike protein, but unlike the main protease, there was no consistency in the use of a common PDB structure from the paper that studied the spike protein.\n",
       "\n",
       "### Methodology\n",
       "\n",
       "A single PDB structure with the highest resolution was chosen for binding site pharmacophore analysis, as the sequences of the PDB structures were similar and thus likely to have similar structures.\n",
       "\n",
       "### Results\n",
       "\n",
       "The pharmacophore analysis failed due to the diversity of the drugs used. **Summary**\n",
       "================\n",
       "\n",
       "**Introduction**\n",
       "---------------\n",
       "\n",
       "The current research on drug repurposing using molecular docking and in silico drug discovery approaches in the fight against COVID-19 is reviewed.\n",
       "\n",
       "**Main Findings**\n",
       "----------------\n",
       "\n",
       "* Antiviral drugs were the most studied drugs.\n",
       "* The main protease was identified as a potential target due to its central role in viral replication and availability.\n",
       "* Pharmacophore investigation showed promising results with tetracyclines and antivirals, which were originally protease inhibitors.\n",
       "\n",
       "**Recommendations**\n",
       "-------------------\n",
       "\n",
       "* The models should be used in further research, including virtual screening, to aid in finding effective therapeutic agents against COVID-19.\n",
       "\n",
       "**Author Contributions**\n",
       "----------------------\n",
       "\n",
       "All authors listed have significantly contributed to the development and writing of this article. **Declarations**\n",
       "================\n",
       "\n",
       "### Conflict of Interest\n",
       "\n",
       "The authors declare no competing interests.\n",
       "\n",
       "### Human and Animal Rights and Informed Consent\n",
       "\n",
       "This article does not contain any studies with human or animal subjects performed by any of the authors."
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Concatenate the text in the 'sammary' column, filtering out None values\n",
    "concatenated_text = ' '.join(filter(None, df['summary'].tolist()))\n",
    "\n",
    "Markdown(concatenated_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>content</th>\n",
       "      <th>citation_count</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td># Molecular Docking As A Potential Approach In...</td>\n",
       "      <td>Molecular Docking As A Potential Approach In R...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Molecular Docking as a Potential Approach in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>## Abstract</td>\n",
       "      <td>Abstract\\n\\nPurpose of Review This article pro...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Summary of Recent Literature on Repurposing ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>## Introduction</td>\n",
       "      <td>Introduction\\n\\nTowards the end of 2019, a nov...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Summary**\\n===============\\n\\n**Introduction...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>## Methods Workfow Search Strategy</td>\n",
       "      <td>Methods Workfow Search Strategy\\n\\nDuring the ...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Methods Workflow Search Strategy**\\n========...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>## Data Selection And Extraction</td>\n",
       "      <td>Data Selection And Extraction\\n\\nTitles and ab...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Data Selection And Extraction**\\n===========...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>## Pharmacophore Analysis</td>\n",
       "      <td>Pharmacophore Analysis\\n\\nTo perform the pharm...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Pharmacophore Analysis**\\n==================...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>## Data Description</td>\n",
       "      <td>Data Description\\n\\nIn silico work on COVID-19...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Summary of In Silico Work on COVID-19 Drug D...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>## Results And Discussion Main Protease</td>\n",
       "      <td>Results And Discussion Main Protease\\n\\nThe fa...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Results and Discussion: Main Protease**\\n===...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>## Antivirals</td>\n",
       "      <td>Antivirals\\n\\nOf the 29 antiviral drugs that w...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Antivirals**\\n================\\n\\n### Introd...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>## Tetracyclines</td>\n",
       "      <td>Tetracyclines\\n\\nAnother class of compounds th...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Tetracyclines**\\n================\\n\\n### Ove...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>## Pharmacophore Analysis Of The Main Protease</td>\n",
       "      <td>Pharmacophore Analysis Of The Main Protease\\n\\...</td>\n",
       "      <td>0</td>\n",
       "      <td>Here is a summarized version of the text in a ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>## RnaDependent Rna Polymerase</td>\n",
       "      <td>RnaDependent Rna Polymerase\\n\\nRNA-dependent ...</td>\n",
       "      <td>0</td>\n",
       "      <td>**RNA-Dependent RNA Polymerase (RdRp)**\\n=====...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>## Drug Afecting RnaDependent Rna Polymerase</td>\n",
       "      <td>Drug Afecting RnaDependent Rna Polymerase\\n\\n...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Drug Afecting Rna-Dependent Rna Polymerase**...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>## Pharmacophore Analysis Of The RnaDependent...</td>\n",
       "      <td>Pharmacophore Analysis Of The RnaDependent Rn...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Pharmacophore Analysis of RNA-Dependent RNA ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>## Spike Protein</td>\n",
       "      <td>Spike Protein\\n\\n![8_Image_1.Png](8_Image_1.Pn...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Spike Protein**\\n================\\n\\n### Ove...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>## Drug Afecting Spike Protein</td>\n",
       "      <td>Drug Afecting Spike Protein\\n\\n| Table 5 Drugs...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Drug Affecting Spike Protein**\\n============...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>## Drug Afecting The Ace2/Spike Complex</td>\n",
       "      <td>Drug Afecting The Ace2/Spike Complex\\n\\nA numb...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Drug Afecting The ACE2/Spike Complex**\\n====...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>## Pharmacophore Analysis Of The Spike Protein</td>\n",
       "      <td>Pharmacophore Analysis Of The Spike Protein\\n\\...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Pharmacophore Analysis of the Spike Protein*...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>## Conclusion</td>\n",
       "      <td>Conclusion\\n\\nHerein, we reviewed the current ...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Summary**\\n================\\n\\n**Introductio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>## Declarations</td>\n",
       "      <td>Declarations\\n\\nConflict of Interest The autho...</td>\n",
       "      <td>0</td>\n",
       "      <td>**Declarations**\\n================\\n\\n### Conf...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>## References</td>\n",
       "      <td>References\\n\\nPapers of particular interest, p...</td>\n",
       "      <td>108</td>\n",
       "      <td>Summary not generated for reference sections.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                title  \\\n",
       "0   # Molecular Docking As A Potential Approach In...   \n",
       "1                                         ## Abstract   \n",
       "2                                     ## Introduction   \n",
       "3                  ## Methods Workfow Search Strategy   \n",
       "4                    ## Data Selection And Extraction   \n",
       "5                           ## Pharmacophore Analysis   \n",
       "6                                 ## Data Description   \n",
       "7             ## Results And Discussion Main Protease   \n",
       "8                                       ## Antivirals   \n",
       "9                                    ## Tetracyclines   \n",
       "10     ## Pharmacophore Analysis Of The Main Protease   \n",
       "11                    ## RnaDependent Rna Polymerase   \n",
       "12      ## Drug Afecting RnaDependent Rna Polymerase   \n",
       "13  ## Pharmacophore Analysis Of The RnaDependent...   \n",
       "14                                   ## Spike Protein   \n",
       "15                     ## Drug Afecting Spike Protein   \n",
       "16            ## Drug Afecting The Ace2/Spike Complex   \n",
       "17     ## Pharmacophore Analysis Of The Spike Protein   \n",
       "18                                      ## Conclusion   \n",
       "19                                    ## Declarations   \n",
       "20                                      ## References   \n",
       "\n",
       "                                              content  citation_count  \\\n",
       "0   Molecular Docking As A Potential Approach In R...               0   \n",
       "1   Abstract\\n\\nPurpose of Review This article pro...               0   \n",
       "2   Introduction\\n\\nTowards the end of 2019, a nov...               0   \n",
       "3   Methods Workfow Search Strategy\\n\\nDuring the ...               0   \n",
       "4   Data Selection And Extraction\\n\\nTitles and ab...               0   \n",
       "5   Pharmacophore Analysis\\n\\nTo perform the pharm...               0   \n",
       "6   Data Description\\n\\nIn silico work on COVID-19...               0   \n",
       "7   Results And Discussion Main Protease\\n\\nThe fa...               0   \n",
       "8   Antivirals\\n\\nOf the 29 antiviral drugs that w...               0   \n",
       "9   Tetracyclines\\n\\nAnother class of compounds th...               0   \n",
       "10  Pharmacophore Analysis Of The Main Protease\\n\\...               0   \n",
       "11  RnaDependent Rna Polymerase\\n\\nRNA-dependent ...               0   \n",
       "12  Drug Afecting RnaDependent Rna Polymerase\\n\\n...               0   \n",
       "13  Pharmacophore Analysis Of The RnaDependent Rn...               0   \n",
       "14  Spike Protein\\n\\n![8_Image_1.Png](8_Image_1.Pn...               0   \n",
       "15  Drug Afecting Spike Protein\\n\\n| Table 5 Drugs...               0   \n",
       "16  Drug Afecting The Ace2/Spike Complex\\n\\nA numb...               0   \n",
       "17  Pharmacophore Analysis Of The Spike Protein\\n\\...               0   \n",
       "18  Conclusion\\n\\nHerein, we reviewed the current ...               0   \n",
       "19  Declarations\\n\\nConflict of Interest The autho...               0   \n",
       "20  References\\n\\nPapers of particular interest, p...             108   \n",
       "\n",
       "                                              summary  \n",
       "0   **Molecular Docking as a Potential Approach in...  \n",
       "1   **Summary of Recent Literature on Repurposing ...  \n",
       "2   **Summary**\\n===============\\n\\n**Introduction...  \n",
       "3   **Methods Workflow Search Strategy**\\n========...  \n",
       "4   **Data Selection And Extraction**\\n===========...  \n",
       "5   **Pharmacophore Analysis**\\n==================...  \n",
       "6   **Summary of In Silico Work on COVID-19 Drug D...  \n",
       "7   **Results and Discussion: Main Protease**\\n===...  \n",
       "8   **Antivirals**\\n================\\n\\n### Introd...  \n",
       "9   **Tetracyclines**\\n================\\n\\n### Ove...  \n",
       "10  Here is a summarized version of the text in a ...  \n",
       "11  **RNA-Dependent RNA Polymerase (RdRp)**\\n=====...  \n",
       "12  **Drug Afecting Rna-Dependent Rna Polymerase**...  \n",
       "13  **Pharmacophore Analysis of RNA-Dependent RNA ...  \n",
       "14  **Spike Protein**\\n================\\n\\n### Ove...  \n",
       "15  **Drug Affecting Spike Protein**\\n============...  \n",
       "16  **Drug Afecting The ACE2/Spike Complex**\\n====...  \n",
       "17  **Pharmacophore Analysis of the Spike Protein*...  \n",
       "18  **Summary**\\n================\\n\\n**Introductio...  \n",
       "19  **Declarations**\\n================\\n\\n### Conf...  \n",
       "20      Summary not generated for reference sections.  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_with_citations_and_summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [
    "IlckXLp5p-7c"
   ],
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
